letsencrypt-auto-source/pieces/check_permissions.py

```diff
@@ -0,0 +1,80 @@
+"""Verifies certbot-auto cannot be modified by unprivileged users.
+
+This script takes the path to certbot-auto as its only command line
+argument.  It then checks that the file can only be modified by uid/gid
+< 1000 and if other users can modify the file, it prints a warning with
+a suggestion on how to solve the problem.
+
+If the absolute path of certbot-auto contains a symlink, it is not
+handled specially and the symlink is followed. Due to this, there could
+be permissions problems unreported by this script, however, issues like
+this were not caused by our documentation and are ignored for the sake
+of simplicity.
+
+All warnings are printed to stdout rather than stderr so all stderr
+output from this script can be suppressed to avoid printing messages if
+this script fails for some reason.
+
+"""
+from __future__ import print_function
+
+import os
+import stat
+import sys
+
+
+FORUM_POST_URL = 'https://community.letsencrypt.org/t/certbot-auto-deployment-best-practices/91979/'
+
+
+def has_safe_permissions(path):
+    """Returns True if the given path has secure permissions.
+
+    The permissions are considered safe if the file is only writable by
+    uid/gid < 1000.
+
+    The reason we allow more IDs than 0 is because on some systems such
+    as Debian, system users/groups other than uid/gid 0 are used for the
+    path we recommend in our instructions which is /usr/local/bin.  1000
+    was chosen because on Debian 0-999 is reserved for system IDs[1] and
+    on RHEL either 0-499 or 0-999 is reserved depending on the
+    version[2][3]. Due to these differences across different OSes, this
+    detection isn't perfect so we only determine permissions are
+    insecure when we can be reasonably confident there is a problem
+    regardless of the underlying OS.
+
+    [1] https://www.debian.org/doc/debian-policy/ch-opersys.html#uid-and-gid-classes
+    [2] https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/deployment_guide/ch-managing_users_and_groups
+    [3] https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/system_administrators_guide/ch-managing_users_and_groups
+
+    :param str path: filesystem path to check
+    :returns: True if the path has secure permissions, otherwise, False
+    :rtype: bool
+
+    """
+    stat_result = os.stat(path)
+    if stat_result.st_mode & stat.S_IWOTH:
+        return False
+    if stat_result.st_mode & stat.S_IWGRP and stat_result.st_gid >= 1000:
+        return False
+    if stat_result.st_mode & stat.S_IWUSR and stat_result.st_uid >= 1000:
+        return False
+    return True
+
+
+def main(certbot_auto_path):
+    current_path = os.path.abspath(certbot_auto_path)
```

Comment:
OK. I tried very hard, and finally found an insane corner case. I am not sure that it worth to study it, but let's expose it.

Let's consider that the real path of certbot-auto is `/root/certbot/certbot-auto`. Let's consider also that we have a symlink `/tmp/certbot` pointing to `/root/certbot` AND some insane admin sys executed the command `chown -h myuser:myuser /tmp/certbot` (or he created the symlink as a normal user, not using sudo ...).

Let's consider now:

1) That certbot-auto is called through `/tmp/certbot/certbot-auto`.
In this case, the input from `"$0"` is `/tmp/certbot/certbot-auto`, and path resolution here will be `/tmp/certbot/certbot-auto`. When checking `/tmp/certbot`, so on the symlink owned by `myuser` pointing to `/root/certbot` owned by `root`, will raise a `PermissionDenied` (even running as root, as the permissions between the symlink and its target are not corresponding.

2) That certbot-auto is called from `/tmp/certbot` using `./certbot-auto`
In this case, the input from `"$0"` is `./certbot-auto` and path resolution is `/root/certbot/certbot-auto` because `abspath` will resolve the symlinks. It means that the permissions will be seen as safe. But `/tmp/certbot` is owned by `myuser`, that can delete the symlink and point it to somewhere else.

In the first situation, I think that we should check if the current path is in fact a symlink, and check its ownership. In the second situation, I do not know how to get the path including symlinks from bash `"$0"` variable.

Reply:
Nice catch! Case (1) is a type of problem I hadn't considered. Luckily the code at https://github.com/certbot/certbot/pull/6995/files#diff-a4881cc6ba1c9dcaeb3962d88a0c369dR1598 seems to properly work to silence all error output and keep certbot-auto from crashing, but this exception certainly isn't the behavior I had in mind.

To fix this, I changed the call to `os.path.abspath` to `os.path.realpath` to eliminate any symlinks in the path upfront to avoid problems like this.

For the permission problem described in (2), I think we should ignore problems like this. I wrote more about this at https://github.com/certbot/certbot/pull/6995/files#diff-932f0067f1e5ad8700ea2c5b1a14a043R8.

Reply:
I agree!